Mathlib/Analysis/PSeries.lean

```diff
@@ -19,134 +19,180 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
-    ENNReal.coe_pow, ENNReal.coe_two]
+/-- for a series of `NNReal` version. -/
+theorem summable_schlomilch_iff {C : ℕ} {u : ℕ → ℕ} {f : ℕ → ℝ≥0}
+    (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m)
+    (h_pos : ∀ n, 0 < u n) (h_nonneg : ∀ n, 0 <= f n) (hu_strict : StrictMono u)
+    (hC_nonzero : C ≠ 0) (h_succ_diff : SuccDiffBounded C u) :
+    (Summable fun k : ℕ => (u (k + 1) - (u k : ℝ≥0)) * f (u k)) ↔ Summable f := by
+  simp only [← tsum_coe_ne_top_iff_summable, Ne.def, not_iff_not, ENNReal.coe_mul]
   constructor <;> intro h
   · replace hf : ∀ m n, 1 < m → m ≤ n → (f n : ℝ≥0∞) ≤ f m := fun m n hm hmn =>
       ENNReal.coe_le_coe.2 (hf (zero_lt_one.trans hm) hmn)
-    simpa [h, ENNReal.add_eq_top, ENNReal.mul_eq_top] using ENNReal.tsum_condensed_le hf
+    replace h_nonneg : ∀ n, 0 ≤ (f n : ℝ≥0∞) := fun n =>
+      ENNReal.coe_le_coe.2 (h_nonneg n)
+    obtain hC := tsum_schlomilch_le hf h_pos h_nonneg hu_strict h_succ_diff
+    have : (↑(u 1) - ↑(u 0)) * ↑(f (u 0)) + ↑C * ∑' (k : ℕ), ↑(f k) = ∞ := by exact eq_top_mono hC h
+    simpa [add_eq_top, mul_ne_top, mul_eq_top, hC_nonzero]
   · replace hf : ∀ m n, 0 < m → m ≤ n → (f n : ℝ≥0∞) ≤ f m := fun m n hm hmn =>
       ENNReal.coe_le_coe.2 (hf hm hmn)
-    simpa [h, ENNReal.add_eq_top] using ENNReal.le_tsum_condensed hf
-#align nnreal.summable_condensed_iff NNReal.summable_condensed_iff
-
+    have : ∑ k in range (u 0), ↑(f k) ≠ ∞ := by
+      apply ne_top_of_lt (sum_lt_top fun a _ => coe_ne_top)
+    simpa [h, add_eq_top, this] using le_tsum_schlomilch hf h_pos hu_strict
 end NNReal
 
 open NNReal in
-/-- Cauchy condensation test for series of nonnegative real numbers. -/
+/-- for series of nonnegative real numbers. -/
+theorem summable_schlomilch_iff_of_nonneg {C : ℕ} {u : ℕ → ℕ} {f : ℕ → ℝ} (h_nonneg : ∀ n, 0 ≤ f n)
+    (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (hu_strict : StrictMono u) (hC_nonzero : C ≠ 0) (h_succ_diff : SuccDiffBounded C u) :
+    (Summable fun k : ℕ => (u (k + 1) - (u k : ℝ)) * f (u k)) ↔ Summable f := by
+  lift f to ℕ → ℝ≥0
+  exact h_nonneg
+  simp only [NNReal.coe_le_coe] at *
+  have (k : ℕ) : (u (k + 1) - (u k : ℝ)) = ((u (k + 1) : ℝ≥0) - (u k : ℝ≥0) : ℝ≥0) := by
+    have := Nat.cast_le (α := ℝ≥0).mpr <| (hu_strict k.lt_succ_self).le
+    simp [NNReal.coe_sub this]
+  simp_rw [this]
+  exact_mod_cast NNReal.summable_schlomilch_iff hf h_pos h_nonneg hu_strict hC_nonzero h_succ_diff
+
 theorem summable_condensed_iff_of_nonneg {f : ℕ → ℝ} (h_nonneg : ∀ n, 0 ≤ f n)
     (h_mono : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) :
     (Summable fun k : ℕ => (2 : ℝ) ^ k * f (2 ^ k)) ↔ Summable f := by
-  lift f to ℕ → ℝ≥0 using h_nonneg
-  simp only [NNReal.coe_le_coe] at *
-  exact_mod_cast NNReal.summable_condensed_iff h_mono
-#align summable_condensed_iff_of_nonneg summable_condensed_iff_of_nonneg
+  have h_pos : ∀ (n : ℕ), 0 < 2 ^ n := fun n => pow_pos zero_lt_two n
```

Comment:
You can also simplify this have, using "nested have's".

Reply:
I must have missed this in the documentation. Can you explain what you mean by nested have's?